### ISSUE TITLE
[Issue]: rms_norm and unquantized fused_moe accumulate in bf16, flipping argmax on near-uniform logits

### ISSUE THREAD

**Issue by okorzh-amd** [ROCm/aiter#3806]
### Problem Description

AITER's rms_norm and the unquantized (QuantType.No) fused_moe accumulate their reductions in bfloat16 rather than upcasting to fp32. Per-call error is small (~1 bf16 ULP for rms_norm, ~0.2% RMS-relative for fused_moe) — irrelevant for a trained model, but enough to flip the top-1 argmax when logits are near-uniform (untrained/randomly-init model, rank1–rank2 gap O(1/√V) ≈ a few milli-nats). It surfaced as a vLLM CI failure: greedy decode of TitanML/tiny-mixtral diverges from HF at the tail token with AITER on; workaround is VLLM_ROCM_USE_AITER_RMSNORM=0 + VLLM_ROCM_USE_AITER_MOE=0.

Environment: MI355X (gfx950), ROCm 7.2.3, PyTorch 2.10, aiter from vllm/vllm-openai-rocm:nightly-6fbfdd18…

Observed output (MI355X):

[rms_norm]   |Δ| ulps max=1.00 mean=0.36  exact=0.0%
[rms_norm]   near-uniform argmax flips: 9/4096 (gap=6.97e-03)
[fused_moe]  convention-check (torch_moe vs fp32 ref) max=5.49e-05
[fused_moe]  |Δ|/rms(out) max=2.19e-02 mean=2.24e-03  exact=0.0%
[fused_moe]  near-uniform argmax flips: 5/512 (gap=6.95e-03)
Expected: reduction accumulators in rms_norm (variance) and fused_moe (GEMM/activation) upcast to fp32, matching the native path, so outputs agree to bf16 round-off of the final result rather than accumulating bf16 error across the reduction.

The full file also contains the inlined standalone repro (with the shuffle_weight step and the torch_moe convention check), the run command, and the per-token→32-step flip-probability reasoning (~32%, matching the intermittent failures).

### Operating System

Ubuntu 22

### CPU

AMD EPYC 9575F 

### GPU

mi355x

### ROCm Version

7.2

### Steps to Reproduce

```
#!/usr/bin/env python3
"""Minimal standalone repro for AITER bf16 accumulation drift.

No vLLM dependency. Shows that AITER's `rms_norm` and unquantized `fused_moe`
diverge from an fp32-accumulation reference by ~1 bf16 ULP, and that this is
enough to flip the top-1 argmax when the downstream logits are near-uniform
(the regime of an *untrained* model such as TitanML/tiny-mixtral, whose
rank1/rank2 logit gap is O(1/sqrt(V)) ~ a few milli-nats).

Run:
    HIP_VISIBLE_DEVICES=0 python3 aiter_drift_repro.py
"""

import torch
import aiter

torch.manual_seed(0)
DEV = "cuda"
DT = torch.bfloat16
EPS = 1e-6


def banner(t):
    print("\n" + "=" * 72 + f"\n{t}\n" + "=" * 72)


def ulp_bf16(x_fp32):
    # magnitude of 1 ULP of a bf16 number (8-bit mantissa) at the given value
    return x_fp32.abs().clamp_min(1e-30) * (2.0**-8)


# ---------------------------------------------------------------------------
# Shared "untrained lm_head": small random projection => near-uniform logits.
# rank1-rank2 gap is intentionally tiny, like a randomly-initialised model.
# ---------------------------------------------------------------------------
def argmax_flips(h_a_fp32, h_b_fp32, vocab=32000, scale=0.03):
    # Row-normalise so the comparison isolates the *relative* hidden-state
    # perturbation, then project with a small random lm_head -> near-uniform
    # logits (small rank1-rank2 gap), mimicking an untrained model.
    H = h_a_fp32.shape[-1]
    ha = h_a_fp32 / h_a_fp32.norm(dim=-1, keepdim=True).clamp_min(1e-9)
    hb = h_b_fp32 / h_b_fp32.norm(dim=-1, keepdim=True).clamp_min(1e-9)
    Wlm = torch.randn(H, vocab, device=DEV, dtype=torch.float32) * scale
    la = ha @ Wlm
    lb = hb @ Wlm
    gap = la.topk(2, dim=-1).values.diff(dim=-1).abs()  # rank1-rank2 gap
    flips = (la.argmax(-1) != lb.argmax(-1)).sum().item()
    return flips, la.shape[0], gap.mean().item()


# ===========================================================================
# 1) rms_norm
# ===========================================================================
banner("1) aiter.rms_norm  vs  fp32-accumulation reference  (bf16 in/out)")

N, H = 4096, 4096  # tokens, hidden (mixtral-like)
x = torch.randn(N, H, device=DEV, dtype=DT)
w = torch.randn(H, device=DEV, dtype=DT)


def ref_rmsnorm_fp32(x, w, eps):
    xf = x.float()
    var = xf.pow(2).mean(-1, keepdim=True)
    return (xf * torch.rsqrt(var + eps)) * w.float()


y_aiter = aiter.rms_norm(x, w, EPS).float()
y_ref = ref_rmsnorm_fp32(x, w, EPS)

d = (y_aiter - y_ref).abs()
ulps = (d / ulp_bf16(y_ref))
print(f"  shape={tuple(x.shape)}  dtype={DT}")
print(f"  |Δ| abs   max={d.max():.3e}  mean={d.mean():.3e}")
print(f"  |Δ| ulps  max={ulps.max():.2f}  mean={ulps.mean():.2f}")
print(f"  exact-match elements: {(d == 0).float().mean()*100:.1f}%")

f, n, gap = argmax_flips(y_aiter, y_ref)
print(f"  near-uniform argmax flips (V=32000): {f}/{n}  (mean rank1-2 gap={gap:.3e})")

# ===========================================================================
# 2) fused_moe (unquantized bf16)  -- mixtral is MoE; this is the path the
#    PR additionally disables (VLLM_ROCM_USE_AITER_MOE=0).
# ===========================================================================
banner("2) aiter.fused_moe (bf16)  vs  fp32-accumulation reference")

from aiter import ActivationType, QuantType
from aiter.fused_moe import fused_moe as aiter_fused_moe
from aiter.fused_moe import torch_moe as aiter_torch_moe
from aiter.ops.shuffle import shuffle_weight

T = 512           # tokens
Hm = 1024         # hidden
Im = 2048         # intermediate
E = 8             # experts (mixtral)
TOPK = 2

hs = torch.randn(T, Hm, device=DEV, dtype=DT) * 0.1
# w1 = [E, 2*I, H] (gate+up for silu),  w2 = [E, H, I]
w1 = torch.randn(E, 2 * Im, Hm, device=DEV, dtype=DT) * (Hm**-0.5)
w2 = torch.randn(E, Hm, Im, device=DEV, dtype=DT) * (Im**-0.5)

# near-uniform router (untrained): tiny gate weights => flat softmax
gate = torch.randn(Hm, E, device=DEV, dtype=torch.float32) * (Hm**-0.5) * 0.05
router_logits = hs.float() @ gate
probs = torch.softmax(router_logits, dim=-1)
topk_weight, topk_ids = torch.topk(probs, TOPK, dim=-1)
topk_weight = (topk_weight / topk_weight.sum(-1, keepdim=True)).to(torch.float32)
topk_ids = topk_ids.to(torch.int32)


def ref_moe_fp32(hs, w1, w2, tw, tid):
    hf = hs.float()
    w1f, w2f = w1.float(), w2.float()
    out = torch.zeros_like(hf)
    for t in range(hf.shape[0]):
        acc = torch.zeros(hf.shape[1], device=DEV, dtype=torch.float32)
        for k in range(tid.shape[1]):
            e = int(tid[t, k])
            g_u = hf[t] @ w1f[e].t()          # [2I]
            g, u = g_u.chunk(2, dim=-1)
            act = torch.nn.functional.silu(g) * u
            acc += float(tw[t, k]) * (act @ w2f[e].t())
        out[t] = acc
    return out


# AITER's fused_moe kernel consumes weights in a tiled layout, so vLLM
# pre-shuffles them at load time (oracle/unquantized.py -> shuffle_weights,
# default layout (16,16)). The fp32 reference uses the *original* weights.
w1_s = shuffle_weight(w1, layout=(16, 16))
w2_s = shuffle_weight(w2, layout=(16, 16))
out_aiter = aiter_fused_moe(
    hs, w1_s, w2_s, topk_weight, topk_ids,
    activation=ActivationType.Silu, quant_type=QuantType.No,
).float()
out_ref = ref_moe_fp32(hs, w1, w2, topk_weight, topk_ids)

# sanity: confirm our fp32 reference matches aiter's own torch reference in
# convention (so the diff below is purely bf16 vs fp32, not a layout mistake).
out_torch = aiter_torch_moe(
    hs, w1, w2, topk_weight, topk_ids, activation=ActivationType.Silu
).float()
conv = (out_torch - out_ref).abs().max().item()
print(f"  ref-convention check |aiter.torch_moe - fp32_ref| max={conv:.3e}")

d2 = (out_aiter - out_ref).abs()
rms_out = out_ref.pow(2).mean().sqrt()
print(f"  tokens={T} hidden={Hm} inter={Im} experts={E} topk={TOPK}")
print(f"  |Δ| abs        max={d2.max():.3e}  mean={d2.mean():.3e}")
print(f"  |Δ|/rms(out)   max={(d2.max()/rms_out):.3e}  mean={(d2.mean()/rms_out):.3e}")
print(f"  exact-match elements: {(d2 == 0).float().mean()*100:.1f}%")
f2, n2, gap2 = argmax_flips(out_aiter, out_ref)
print(f"  near-uniform argmax flips (V=32000): {f2}/{n2}  (mean rank1-2 gap={gap2:.3e})")

print("\nDONE")
```

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

### Additional Information

follow up for Vllm PR https://github.com/vllm-project/vllm/pull/45509
